sql-statements/sql-statement-alter-index.md

```diff
@@ -0,0 +1,204 @@
+| Sort_4                  | 10000.00 | root      |               | test.t1.c1:asc                 |
+| └─TableReader_8         | 10000.00 | root      |               | data:TableFullScan_7           |
+|   └─TableFullScan_7     | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo |
++-------------------------+----------+-----------+---------------+--------------------------------+
+3 rows in set (0.00 sec)
+```
+
+By comparison, `c2` is a **visible index** and can be used by the optimizer.
+
+{{< copyable "sql" >}}
+
+```sql
+EXPLAIN SELECT c2 FROM t1 ORDER BY c2;
+```
+
+```sql
++------------------------+----------+-----------+------------------------+-------------------------------+
+| id                     | estRows  | task      | access object          | operator info                 |
++------------------------+----------+-----------+------------------------+-------------------------------+
+| IndexReader_13         | 10000.00 | root      |                        | index:IndexFullScan_12        |
+| └─IndexFullScan_12     | 10000.00 | cop[tikv] | table:t1, index:c2(c2) | keep order:true, stats:pseudo |
++------------------------+----------+-----------+------------------------+-------------------------------+
+2 rows in set (0.00 sec)
+```
+
+Even if you use the `USE INDEX` SQL hint to forcibly use indexes, the optimizer still cannot use invisible indexes; otherwise, an error is returned.
+
+{{< copyable "sql" >}}
+
+```sql
+SELECT * FROM t1 USE INDEX(c1);
+```
+
+```sql
+ERROR 1176 (42000): Key 'c1' doesn't exist in table 't1'
+```
+
+> **Note:**
+>
+> "Invisible" here means invisible only to the optimizer. You can still modify or delete invisible indexes.
+
+{{< copyable "sql" >}}
+
+```sql
+ALTER TABLE t1 DROP INDEX c1;
+```
+
+```sql
+Query OK, 0 rows affected (0.02 sec)
+```
+
+### Restriction
+
+There is a restriction on invisible indexes in MySQL: the **primary key** cannot be set to `Invisible`. TiDB is compatible with this restriction. If you set the primary key to `Invisible` in TiDB, an error is returned.
+
+{{< copyable "sql" >}}
+
+```sql
+CREATE TABLE t2(c1 INT, PRIMARY KEY(c1) INVISIBLE);
+```
+
+```sql
+ERROR 3522 (HY000): A primary key index cannot be invisible
+```
+
+The **primary key** above includes both the explicit primary key (specified using the `PRIMARY KEY` keyword) and the implicit primary key.
+
+If the explicit primary key does not exist in the table, the first `UNIQUE` index (every column of the index must be `NOT NULL`) automatically becomes the implicit index.
+
+This means that you cannot set this implicit primary key to `Invisible`.
+
+> **Note:**
+>
+> TiDB does not actually create an **implicit primary key**. TiDB is compatible with this MySQL restriction only in behaviors.
+
+{{< copyable "sql" >}}
+
+```sql
+CREATE TABLE t2(c1 INT NOT NULL, UNIQUE(c1) INVISIBLE);
+```
+
+```
+ERROR 3522 (HY000): A primary key index cannot be invisible
+```
+
+## See also
```

Comment:
There is a MySQL compatibility difference in TiDB that should be documented. Because TiDB does not support the `optimizer_switch`, the following test case would allow the index to be used in MySQL - but will not work in TiDB:

```sql
DROP TABLE IF EXISTS t1;
CREATE TABLE t1 (
 id INT NOT NULL PRIMARY KEY auto_increment,
 b INT NOT NULL,
 INDEX (b) INVISIBLE
);

INSERT INTO t1 VALUES (1, 1), (2,2),(3,3),(4,4),(5,5);
SELECT SLEEP(1);
ANALYZE TABLE t1;
SET optimizer_switch="use_invisible_indexes=on";
EXPLAIN SELECT * FROM t1 WHERE b = 3;
```

Reply:
PTAL @zimulala

Reply:
Thanks, @wwar, we'll document this difference.